planner/core/plan_cache.go

```diff
@@ -70,10 +71,29 @@ func GetPlanFromSessionPlanCache(ctx context.Context, sctx sessionctx.Context, i
 		}
 	}
 
-	var varsNum int
-	var binVarTypes []byte
-	var txtVarTypes []*types.FieldType
 	isBinProtocol := len(binProtoVars) > 0
+	varsNum, binVarTypes, txtVarTypes := getParamValue(sctx, isBinProtocol, binProtoVars, txtProtoVars)
+
+	if prepared.UseCache && prepared.CachedPlan != nil && !ignorePlanCache { // for point query plan
+		if plan, names, ok, err := getPointQueryPlan(prepared, sessVars, stmtCtx); ok {
+			return plan, names, err
+		}
+	}
+
+	if prepared.UseCache && !ignorePlanCache { // for general plans
+		if plan, names, ok, err := getGeneralPlan(ctx, sctx, cacheKey, bindSQL, is, preparedStmt,
+			binVarTypes, txtVarTypes); err != nil || ok {
+			return plan, names, err
+		}
+	}
+
+	return rebuildPhysicalPlan(ctx, sctx, is, preparedStmt, ignorePlanCache, cacheKey,
+		latestSchemaVersion, isBinProtocol, varsNum, binVarTypes, txtVarTypes)
+}
+
+// getParamValue get real values for "?" in PREPARE statements
+func getParamValue(sctx sessionctx.Context, isBinProtocol bool, binProtoVars []types.Datum,
```

Comment:
How about renaming it to `parseParamTypes`?

Reply:
> How about renaming it to `parseParamTypes`?

Oops，it's indeed `type` not `value`